From a6ef9d94a82cfa5b1c0dd732a987cd45654579e6 Mon Sep 17 00:00:00 2001
From: Hans de Goede <hdegoede@redhat.com>
Date: Wed, 7 Dec 2022 21:26:02 +0100
Subject: [PATCH 2/2] ipu-isys-queue: Fix NULL pointer deref with kernels >=
 6.2

Kernel >= 6.2 set vb2_queue->streaming a bit earlier as <= 6.1,
specifically vb2_queue->streaming is set before the videobuf2-core
enqueues the buffers (just before calling start_streaming).

__buf_queue() uses vb2_queue->streaming to differentiate between
buffers queued before and after start_streaming has been called and
that will now no longer work, leading to a NULL pointer deref of
the not yet setup media_pipe pointer.

Fix this by explicitly checking for media_pipe == NULL. Note this
check cannot be used to replace the vb2_queue->streaming check since
for kernels < 6.1 media_pipe is always set.

Signed-off-by: Hans de Goede <hdegoede@redhat.com>
---
 drivers/media/pci/intel/ipu-isys-queue.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/media/pci/intel/ipu-isys-queue.c b/drivers/media/pci/intel/ipu-isys-queue.c
index a4deef08c..4fd7c32ca 100644
--- a/drivers/media/pci/intel/ipu-isys-queue.c
+++ b/drivers/media/pci/intel/ipu-isys-queue.c
@@ -566,7 +566,7 @@ static void __buf_queue(struct vb2_buffer *vb, bool force)
 	if (ib->req)
 		return;
 
-	if (!pipe_av || !vb->vb2_queue->streaming) {
+	if (!pipe_av || !media_pipe || !vb->vb2_queue->streaming) {
 		dev_dbg(&av->isys->adev->dev,
 			"not pipe_av set, adding to incoming\n");
 		return;
-- 
2.39.0

